{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#install these 5 libraries\n",
    "library(nnls) \n",
    "library(HiClimR)\n",
    "library(ccaPP)\n",
    "library(parallel)\n",
    "library(hdf5r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_in = paste('Discovery_Cohort_Counts.h5ad',sep=\"\") #read in h5ad generated through scanpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize modified CytoTRACE function, saves memory by not outputting a normalized matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#' @title CytoTRACE\n",
    "#'\n",
    "#' @description This function generates single-cell predictions of differentiation status from scRNA-seq data.\n",
    "#' It takes in a matrix of gene expression values where columns are cells and rows are genes.\\cr\\cr\n",
    "#' (Optional) Technical batches can be corrected with ComBat by inputting a character vector of batch\n",
    "#' annotations matching the number of columns (cells) in the matrix. Furthermore, for the analysis of\n",
    "#' large datasets (>3,000 cells), users can increase the speed performance by enabling a subsampling approach for calculation\n",
    "#' and using multiple cores.\n",
    "#'\n",
    "#' @param mat matrix of gene expression values where columns are cells and rows are genes\n",
    "#' @param batch character vector of length equal to the number of columns (cells) in the matrix\n",
    "#' @param enableFast boolean indicating whether or not to run CytoTRACE in 'fast mode' for datasets with >3,000 cells. Fast mode uses a subsampling approach to reduce runtime.\n",
    "#' @param ncores integer indicating the number of cores to utilize when enableFast = TRUE\n",
    "#' @param subsamplesize integer indicating the number of cells to subsample when enableFast = TRUE\n",
    "#'\n",
    "#' @return a list containing\n",
    "#' \\itemize{\n",
    "#' \\item CytoTRACE: a numeric vector of the predicted ordering of single cells from 1.0 (least differentiated) to 0.0 (most differentiated)\n",
    "#' \\item CytoTRACErank: a numeric vector of the ranked predicted ordering of single cells by differentiation status. High ranks correspond to less differentiated cells, while low ranks correspond to more differentiated cells.\n",
    "#' \\item cytoGenes: a numeric vector of the Pearson correlation of each gene with CytoTRACE, ordered from highest to lowest\n",
    "#' \\item GCS: a numeric vector of the gene counts signature (geometric mean of the top 200 genes associated with gene counts)\n",
    "#' \\item gcsGenes: a numeric vector of the Pearson correlation of each gene with gene counts, ordered from highest to lowest\n",
    "#' \\item Counts: a numeric vector of the number of genes expressed per single cell (gene counts)\n",
    "#' \\item filteredCells: a character vector of the names of single cells (columns) that were filtered due to poor quality.\n",
    "#' \\item exprMatrix: a matrix of gene expression values after the following normalization steps: (i) sequencing depth normalization by rescaling single-cell transcriptomes to transcripts per million (TPM) or counts per million (CPM) (ii) log2-normalization with a pseudo-count of 1, and (iii) Census normalization (Qiu et al., 2017) to convert the gene expression matrix to relative transcript counts by rescaling single-cell transcriptomes to the total number of detectably expressed genes in that cell.\n",
    "#'\n",
    "#' }\n",
    "#'\n",
    "#' @author Gunsagar Gulati <cytotrace@gmail.com>\n",
    "#'\n",
    "#' @seealso https://cytotrace.stanford.edu\n",
    "#'\n",
    "#' @references https://doi.org/10.1101/649848\n",
    "#'\n",
    "#' @examples\n",
    "#' #Use the bone marrow 10x scRNA-seq dataset to run CytoTRACE\n",
    "#' results <- CytoTRACE(marrow_10x_expr)\n",
    "#'\n",
    "#' #Run this dataset on fast mode using 1 core\n",
    "#' results <- CytoTRACE(marrow_10x_expr, enableFast = TRUE, ncores = 1)\n",
    "#'\n",
    "#' @export\n",
    "\n",
    "CytoTRACE_fixed <- function(mat, batch = NULL, enableFast = TRUE,\n",
    "                      ncores = 1,subsamplesize = 1000){\n",
    "\n",
    "  range01 <- function(x){(x-min(x))/(max(x)-min(x))}\n",
    "\n",
    "  #inputs\n",
    "  a1 <- mat\n",
    "  a2 <- batch\n",
    "  if(ncol(mat) < 3000){\n",
    "    enableFast = FALSE\n",
    "    message(\"The number of cells in your dataset is less than 3,000. Fast mode has been disabled.\")\n",
    "    } else {\n",
    "  message(\"The number of cells in your dataset exceeds 3,000. CytoTRACE will now be run in fast mode (see documentation). You can multi-thread this run using the 'ncores' flag. To disable fast mode, please indicate 'enableFast = FALSE'.\")\n",
    "  }\n",
    "  #Checkpoint: NAs and poor quality genes\n",
    "  pqgenes <- is.na(rowSums(mat>0)) | apply(mat, 1, var) == 0\n",
    "  num_pqgenes <- length(which(pqgenes == TRUE))\n",
    "  mat <- mat[!pqgenes,]\n",
    "  if(num_pqgenes>0){\n",
    "    warning(paste(num_pqgenes, \"genes have zero expression in the matrix and were filtered\"))\n",
    "  }\n",
    "\n",
    "  #Subsample routine\n",
    "  if(enableFast == FALSE){\n",
    "    size <- ncol(mat)\n",
    "  } else if (enableFast == TRUE & subsamplesize < ncol(mat)){\n",
    "    size <- subsamplesize\n",
    "  } else if (enableFast == TRUE & subsamplesize >= ncol(mat)){\n",
    "    stop(\"Please choose a subsample size less than the number of cells in dataset.\")\n",
    "  }\n",
    "\n",
    "  chunk <- round(ncol(mat)/size)\n",
    "  subsamples <- split(1:ncol(mat), sample(factor(1:ncol(mat) %% chunk)))\n",
    "  message(paste(\"CytoTRACE will be run on\", chunk, \"sub-sample(s) of approximately\",\n",
    "                round(mean(unlist(lapply(subsamples, length)))), \"cells each using\", min(chunk, ncores),\"/\", ncores, \"core(s)\"))\n",
    "\n",
    "  message(paste(\"Pre-processing data and generating similarity matrix...\"))\n",
    "  batches <- parallel::mclapply(subsamples, mc.cores = min(chunk, ncores), function(subsample){\n",
    "    #Checkpoint: log2-normalization\n",
    "    mat <- mat[,subsample]\n",
    "    batch <- batch[subsample]\n",
    "\n",
    "    if(max(mat)<50){\n",
    "      mat <- 2^mat - 1\n",
    "    }\n",
    "\n",
    "    #Checkpoint: ERCC standards\n",
    "    if(length(grep(\"ERCC-\", rownames(mat)))>0){\n",
    "      mat <- mat[-grep(\"ERCC-\", rownames(mat)),]\n",
    "    }\n",
    "\n",
    "    #Checkpoint: Sequencing depth normalization\n",
    "    mat <- t(t(mat)/apply(mat, 2, sum))*1000000\n",
    "\n",
    "    #Checkpoint: NAs and poor quality cells\n",
    "    pqcells <- is.na(apply(mat>0, 2, sum)) | apply(mat>0, 2, sum) <= 10\n",
    "    num_pqcells <- length(which(pqcells == TRUE))\n",
    "    mat <- mat[,!pqcells]\n",
    "\n",
    "    #Checkpoint: log2-normalize\n",
    "    mat <- log(mat+1,2)\n",
    "    mat <- data.matrix(mat)\n",
    "\n",
    "    #Calculate pre-batch corrected gene counts\n",
    "    counts <- apply(mat>0, 2, sum)\n",
    "    #Checkpoint: Batch correction\n",
    "    if(ncol(a1) == length(a2)){\n",
    "      #filter poor quality cells from batch vector\n",
    "      batch <- batch[!pqcells]\n",
    "\n",
    "      #Run Combat\n",
    "      suppressMessages(mat <- sva::ComBat(mat, batch, c()))\n",
    "      mat <- data.matrix(mat)\n",
    "\n",
    "      #Replace negative values after batch correction with zeroes for compatibility with downstream steps\n",
    "      mat[which(mat<0)] <- 0\n",
    "    }\n",
    "    #Rescale each single cell with gene counts to convert relative transcript abundances to absolute RNA content prior to cell lysis (credit: Census, Qiu et al., 2017)\n",
    "    census_normalize <- function(mat, counts) {\n",
    "      xnl <- 2^data.matrix(mat) - 1\n",
    "      rs <- apply(xnl, 2, sum)\n",
    "      rnorm <- t(t(xnl) * counts/rs)\n",
    "      A <- log(rnorm+1,2)\n",
    "      return(A)\n",
    "    }\n",
    "\n",
    "    mat2 <- census_normalize(mat, counts)\n",
    "    #Function to identify the most variable genes\n",
    "    mvg <- function(matn) {\n",
    "      A <- matn\n",
    "      n_expr <- rowSums(A > 0);\n",
    "      A_filt <- A[n_expr >= 0.05 * ncol(A),];\n",
    "      vars <- apply(A_filt, 1, var);\n",
    "      means <- apply(A_filt, 1, mean);\n",
    "      disp <- vars / means;\n",
    "      last_disp <- tail(sort(disp), 1000)[1];\n",
    "      A_filt <- A_filt[disp >= last_disp,];\n",
    "\n",
    "      return(A_filt)\n",
    "    }\n",
    "\n",
    "    #Filter out cells not expressing any of the 1000 most variable genes\n",
    "    mat2.mvg <- mvg(mat2)\n",
    "    rm1 <- colSums(mat2.mvg) == 0\n",
    "    mat2 <- mat2[, !rm1]\n",
    "    counts <- counts[!rm1]\n",
    "\n",
    "    #Calculate similarity matrix\n",
    "    similarity_matrix_cleaned <- function(similarity_matrix){\n",
    "      D <- similarity_matrix\n",
    "      cutoff <- mean(as.vector(D))\n",
    "      diag(D) <- 0;\n",
    "      D[which(D < 0)] <- 0;\n",
    "      D[which(D <= cutoff)] <- 0;\n",
    "      Ds <- D\n",
    "      D <- D / rowSums(D);\n",
    "      D[which(rowSums(Ds)==0),] <- 0\n",
    "      return(D)\n",
    "    }\n",
    "    D <- similarity_matrix_cleaned(HiClimR::fastCor(mvg(mat2)))\n",
    "\n",
    "    return(list(mat2 = mat2,counts = counts, D = D))\n",
    "  }\n",
    "  )\n",
    "  #Prepare for downstream steps\n",
    "  mat2 <- do.call(cbind, lapply(batches, function(x) x$mat2))\n",
    "  counts <- do.call(c, lapply(batches, function(x) x$counts))\n",
    "  filter <- colnames(a1)[-which(colnames(a1) %in% colnames(mat2))]\n",
    "  if(length(filter)>0){\n",
    "    warning(paste(length(filter), \"poor quality cells were filtered based on low or no expression. See 'filteredCells' in returned object for names of filtered cells.\"))\n",
    "  }\n",
    "  #Calculate gene counts signature (GCS) or the genes most correlated with gene counts\n",
    "  message(\"Calculating gene counts signature...\")\n",
    "  ds2 <- sapply(1:nrow(mat2), function(x) ccaPP::corPearson(mat2[x,],counts))\n",
    "  names(ds2) <- rownames(mat2)\n",
    "  gcs <- apply(mat2[which(rownames(mat2) %in% names(rev(sort(ds2))[1:200])),],2,mean)\n",
    "\n",
    "  samplesize <- unlist(lapply(lapply(batches, function(x) x$counts), length))\n",
    "  gcs2 <- split(gcs, as.numeric(rep(names(samplesize), samplesize)))\n",
    "  D2 <- lapply(batches, function(x) x$D)\n",
    "\n",
    "  #Regress gene counts signature (GCS) onto similarity matrix\n",
    "  regressed <- function(similarity_matrix_cleaned, score){\n",
    "    out <- nnls::nnls(similarity_matrix_cleaned,score)\n",
    "    score_regressed <- similarity_matrix_cleaned %*% out$x\n",
    "    return(score_regressed)\n",
    "  }\n",
    "\n",
    "  #Apply diffusion to regressed GCS using similarity matrix\n",
    "  diffused <- function(similarity_matrix_cleaned, score, ALPHA = 0.9){\n",
    "    vals <- score\n",
    "    v_prev <- rep(vals);\n",
    "    v_curr <- rep(vals);\n",
    "\n",
    "    for(i in 1:10000) {\n",
    "      v_prev <- rep(v_curr);\n",
    "      v_curr <- ALPHA * (similarity_matrix_cleaned %*% v_curr) + (1 - ALPHA) * vals;\n",
    "\n",
    "      diff <- mean(abs(v_curr - v_prev));\n",
    "      if(diff <= 1e-6) {\n",
    "        break;\n",
    "      }\n",
    "    }\n",
    "    return(v_curr)\n",
    "  }\n",
    "\n",
    "  message(\"Smoothing values with NNLS regression and diffusion...\")\n",
    "  cytotrace <- parallel::mclapply(1:length(D2), mc.cores = ncores, function(i) {\n",
    "    gcs_regressed <- regressed(D2[[i]], gcs2[[i]])\n",
    "    gcs_diffused <- diffused(D2[[i]], gcs_regressed)\n",
    "    cytotrace <- rank(gcs_diffused)\n",
    "  }\n",
    "  )\n",
    "\n",
    "  cytotrace <- cytotrace_ranked <- unlist(cytotrace)\n",
    "  cytotrace <- range01(cytotrace)\n",
    "\n",
    "  #Calculate genes associated with CytoTRACE\n",
    "  cytogenes <- sapply(1:nrow(mat2),\n",
    "                          function(x) ccaPP::corPearson(mat2[x,], cytotrace))\n",
    "  names(cytogenes) <- rownames(mat2)\n",
    "  message(\"Calculating genes associated with CytoTRACE...\")\n",
    "\n",
    "  #Final steps\n",
    "  message(\"Before Final steps\")\n",
    "  names(cytotrace) <- names(cytotrace_ranked) <- names(gcs) <- names(counts) <- colnames(mat2)\n",
    "  cytotrace <- cytotrace[colnames(a1)]; cytotrace_ranked <- cytotrace_ranked[colnames(a1)]; gcs <- gcs[colnames(a1)]; counts <- counts[colnames(a1)]\n",
    "  message(\"Final steps\")\n",
    "                      \n",
    "  #mat2 <- t(data.frame(t(mat2))[colnames(a1),])\n",
    "  message(\"Transposition thing skipped\")                    \n",
    "  names(cytotrace) <- names(cytotrace_ranked) <- names(gcs) <- names(counts) <- colnames(a1) #<- colnames(mat2) \n",
    "\n",
    "  message(\"Done\")\n",
    "  return(list(CytoTRACE = cytotrace, CytoTRACErank = cytotrace_ranked, cytoGenes = sort(cytogenes, decreasing = T), GCS = gcs, gcsGenes = sort(ds2, decreasing = T),\n",
    "              Counts = counts, filteredCells = filter)) #, exprMatrix = mat2\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run this next block of code to define the data read-in function, you may need to replace the '_index' with 'Cells' or 'Genes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function uses hdf5r to interface with the .h5ad filetype output from scanpy\n",
    "prepare_data <- function(h5ad_in) {\n",
    "    data = h5file(h5ad_in)\n",
    "    data_x = data[['X']][,]\n",
    "    data_obs = data[['obs']][['_index']][] #Cells / _index, sometimes these attributes are saved as _index or Cells, replace as needed\n",
    "    data_var = data[['var']][['_index']][] #Genes / _index, sometimes these attributes are saved as _index or Genes, replace as needed\n",
    "    rownames(data_x) = data_var #set labels\n",
    "    colnames(data_x) = data_obs\n",
    "    return(data_x)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run CytoTRACE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The number of cells in your dataset exceeds 3,000. CytoTRACE will now be run in fast mode (see documentation). You can multi-thread this run using the 'ncores' flag. To disable fast mode, please indicate 'enableFast = FALSE'.\n",
      "\n",
      "Warning message in CytoTRACE_fixed(prepare_data(data_in), subsamplesize = 3000, :\n",
      "“10711 genes have zero expression in the matrix and were filtered”\n",
      "CytoTRACE will be run on 20 sub-sample(s) of approximately 3033 cells each using 2 / 2 core(s)\n",
      "\n",
      "Pre-processing data and generating similarity matrix...\n",
      "\n",
      "Warning message in parallel::mclapply(subsamples, mc.cores = min(chunk, ncores), :\n",
      "“scheduled core 2 did not deliver a result, all values of the job will be affected”\n",
      "Warning message in CytoTRACE_fixed(prepare_data(data_in), subsamplesize = 3000, :\n",
      "“30333 poor quality cells were filtered based on low or no expression. See 'filteredCells' in returned object for names of filtered cells.”\n",
      "Calculating gene counts signature...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "results <- CytoTRACE_fixed(prepare_data(data_in), subsamplesize = 3000 ,ncores = 2,enableFast = TRUE) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write 6 separate csvs for the cytotrace outputs, these will correspond to the barcodes used in the input h5ad file.\n",
    "write.csv(results$cytoGenes,'cytoGenes.csv')\n",
    "write.csv(results$CytoTRACErank,'CytoTRACErank.csv')\n",
    "write.csv(results$CytoTRACE,'CytoTRACE.csv')\n",
    "write.csv(results$GCS,'GCS.csv')\n",
    "write.csv(results$Counts,'Counts.csv')\n",
    "write.csv(results$gcsGenes,'gcsGenes.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
